File: frontend/src/app/core/direcciones/direcciones.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { BusquedaDireccionOut, DireccionIn, DireccionOut } from './direcciones.models';

@Injectable({ providedIn: 'root' })
export class DireccionesService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/direcciones`;

  listar(): Observable<DireccionOut[]> {
    return this.http.get<DireccionOut[]>(this.base);
  }

  crear(datos: DireccionIn): Observable<DireccionOut> {
    return this.http.post<DireccionOut>(this.base, datos);
  }

  actualizar(id: string, datos: DireccionIn): Observable<DireccionOut> {
    return this.http.put<DireccionOut>(`${this.base}/${id}`, datos);
  }

  marcarPrincipal(id: string): Observable<DireccionOut> {
    return this.http.post<DireccionOut>(`${this.base}/${id}/principal`, {});
  }

  eliminar(id: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }

  /** Geocodificacion directa: de texto a coordenadas (Pelias de openrouteservice). */
  buscar(texto: string): Observable<BusquedaDireccionOut> {
    return this.http.get<BusquedaDireccionOut>(`${this.base}/buscar`, {
      params: new HttpParams().set('texto', texto),
    });
  }

  /** Geocodificacion inversa: que direccion hay donde se solto el pin del mapa. */
  inversa(latitud: number, longitud: number): Observable<BusquedaDireccionOut> {
    return this.http.get<BusquedaDireccionOut>(`${this.base}/inversa`, {
      params: new HttpParams().set('latitud', latitud).set('longitud', longitud),
    });
  }
}

File: movil/lib/core/direcciones/direcciones_service.dart
import '../api.dart';
import 'direcciones_models.dart';

/// CU20 - libreta de direcciones de la clienta (`/direcciones`).
class DireccionesService {
  Future<List<DireccionOut>> listar() async {
    final respuesta = await api.get('/direcciones');
    return comoLista(respuesta).map(DireccionOut.desdeJson).toList();
  }

  Future<DireccionOut> crear(DireccionIn datos) async {
    final respuesta = await api.post('/direcciones', cuerpo: datos.aJson());
    return DireccionOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<DireccionOut> actualizar(String id, DireccionIn datos) async {
    final respuesta = await api.put('/direcciones/$id', cuerpo: datos.aJson());
    return DireccionOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<DireccionOut> marcarPrincipal(String id) async {
    final respuesta = await api.post('/direcciones/$id/principal');
    return DireccionOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<void> eliminar(String id) => api.delete('/direcciones/$id');

  /// Geocodificacion directa (Pelias de openrouteservice), acotada a Bolivia.
  Future<BusquedaDireccion> buscar(String texto) async {
    final respuesta = await api.get('/direcciones/buscar', query: {'texto': texto});
    return BusquedaDireccion.desdeJson(respuesta as Map<String, dynamic>);
  }

  /// Geocodificacion inversa: que direccion hay donde quedo el pin.
  Future<BusquedaDireccion> inversa(double latitud, double longitud) async {
    final respuesta = await api.get(
      '/direcciones/inversa',
      query: {'latitud': latitud, 'longitud': longitud},
    );
    return BusquedaDireccion.desdeJson(respuesta as Map<String, dynamic>);
  }
}

final direccionesService = DireccionesService();
